### PR TITLE
find_nodes_in_area: Extend maximal count to U32_MAX

### LINKED ISSUE #5275: find_nodes_in_area doesn't work in register_on_generated
I tried a small code

```
minetest.register_on_generated(function(minp, maxp)
    local nodes, count = minetest.find_nodes_in_area(minp, maxp, {"air"})
    print(dump(count))
end)
```

I would expect to get 512000 in clean air, but I get less than 60000, even in clean air. Adding "ignore" doesn't change it. Get ignore = 0

Is the area to large or is it not possible to use find_nodes_in_area in register_on_generated?

### PR DESCRIPTION
**Pull request by SmallJoker**
Improve documentation and code style

Fixes #5275, non-trivial

### CODE REVIEW

**Comment by paramat** [comment]
Not trivial.

**Comment by paramat** [comment]
This will need a limit imposed that throws an error before any tables are added to. Set to a little more than a standard mapchunk, 512000.

**Comment by paramat** [comment]
2^19 = 524288 could be a good limit.

**Comment by paramat** [comment]
Possibility of high memory usage when storing 512000 v3s16s in a Lua table, garbage collection may be slow for these tables. Need to test memory use using this looking for air in empty mapchunks in 'on generated'.

**Comment by MarkuBu** [comment]
How about the option just to count the nodes without generating a table? Sometimes you may only need the count, but not the position. Something like "return_table = false"

**Comment by paramat** [comment]
Interesting idea.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2236,9 +2236,13 @@ and `minetest.auth_reload` call the authetification handler.
 * `minetest.find_nodes_in_area(minp, maxp, nodenames)`: returns a list of positions
     * returns as second value a table with the count of the individual nodes found
```
Comment (paramat) [inline doc/lua_api.txt]: Remove old duplicated description of return values? (below too).

**Comment by paramat** [comment]
Nice idea for the limit.

**Comment by paramat** [comment]
Soory accidental close.

File: doc/lua_api.txt (hunk 1 above)
Comment (SmallJoker) [inline doc/lua_api.txt]: Removed

**Comment by sorcerykid** [comment]
I'm concerned about the limit being raised. Processing an array of several hundred thousand positions in Lua at runtime seems excessive, not to mention impractical. For any mod to be working with data-sets that large should be a huge red flag that something is being done the wrong way :-/

Moreover, I've rarely encountered a situation where I need to know both the positions and the counts of nodes in an area simultaneously. And even then I'm only concerned with the sum of all positions (which can be obtained from the size of the array), so the second return value proves to be redundant.

For this reason, I'm already using a placeholder minetest.count_nodes_in_area( ) function in three of my mods. It has the added benefit that it can translate the node counts back into group counts (so if you search for "group:liquid", then the group is included in the results). I eventually hope to port it to CPP.

```
-----------------------------------------------------------------
-- count number of nodes by name or group within given area
-----------------------------------------------------------------

minetest.count_nodes_in_area = function( pos_min, pos_max, names, is_group )
        local counts, node_counts
        node_counts = select( 2, minetest.find_nodes_in_area( pos_min, pos_max, names ) )

        if is_group == false then
                return node_counts
        end

        counts = { }    -- use new table for transposing node counts into group counts
        for _, name in ipairs( names ) do
                local group_name = string.match( name, "group:(.+)" )
                if group_name then
                        counts[ name ] = 0
                        for node_name, node_count in pairs( node_counts ) do
                                if minetest.registered_nodes[ node_name ].groups[ group_name ] then
                                        counts[ name ] = counts[ name ] + node_count
                                end
                        end
                else
                        counts[ name ] = node_counts[ name ]
                end
        end
        return counts
end
```

**Comment by MarkuBu** [comment]
Did you ever use find_nodes_in_area? I don't think so.

Count returns more than just a single value

**Comment by sorcerykid** [comment]
I'm not sure what you are asking. I use minetest.find_nodes_in_area( ) to search for multiple nodes in a given area, and to count the nodes respectively. But sometimes I want to search by group, rather the node, so hence I created a function that allows for group searches.

**Comment by paramat** [comment]
If we decide against u32 the cleanup and limit here will still be useful.
I'm also concerned about the memory use, perhaps we should not encourage the use of this for very large volumes (> 40^3 nodes).

**Comment by MarkuBu** [comment]
> And even then I'm only concerned with the sum of all positions (which can be obtained from the size of the array), so the second return value proves to be redundant.

You can search for more than one node or for groups of nodes and count returns values for each node. Then you can't obtain the count from the size of the array

**Comment by MarkuBu** [comment]
@sorcerykid this is the result of count if you search for "group:stone". Your post sounds like you didn't know that

```
{
	["default:stone"] = 17812,
	["default:desert_stone_block"] = 0,
	["walls:desertcobble"] = 0,
	["default:mossycobble"] = 0,
	["default:desert_stone"] = 14192,
	["walls:mossycobble"] = 0,
	["default:desert_stonebrick"] = 0,
	["walls:cobble"] = 0,
	["default:cobble"] = 0,
	["default:stonebrick"] = 0,
	["default:desert_cobble"] = 0,
	["default:stone_block"] = 0
}
```

I always think the modder is responsible for the mod performance. I still think, u32 is a good idea, but with a warning. With my propose to disable the table output it can be useful to count nodes in a chunk

**Comment by Fixer-007** [comment]
I hope I will not get even more OOMs from this, it is already ridiculous (at least on 64bit win7sp1)

**Comment by sorcerykid** [comment]
I don't know what you are going on about MarkuBu. If I want the count of the individual nodes, then I use the second return value. If I want the positions I use the first return value. 

> @sorcerykid this is the result of count if you search for "group:stone". Your post sounds like you didn't know that

Huh? If I search for "group:stone" then I expect a count of group stone in the results, not a count of default:stone, default:desert_stone_block, walls:desertcobble, default:mossycobble, default:desert_stone, walls:mossycobble, default:desert_stonebrick, walls:cobble, default:cobble, default:stonebrick, and default:stone_block that I have to then go through the hassle of manually adding up just to get the actual group count that I wanted in the first place.

**Comment by MarkuBu** [comment]
> If I search for "group:stone" then I expect a count of group:stone in the results

Ah, here we go. Well, then you can still use #nodes for counting the total of group:stone. Or maybe we talk past each other

File: src/script/lua_api/l_env.cpp [hunk 2]
```diff
@@ -635,35 +635,46 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
...
 			// removes value, keeps key for next iteration
 			lua_pop(L, 1);
 		}
-	} else if(lua_isstring(L, 3)) {
+	} else if (lua_isstring(L, 3)) {
 		ndef->getIds(lua_tostring(L, 3), filter);
 	}
 
-	std::map<content_t, u16> individual_count;
+	std::map<content_t, u32> individual_count;
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: it can be a unordered map here

File: src/script/lua_api/l_env.cpp [hunk 3]
```diff
@@ -635,35 +635,46 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
...
+	 * and give tolerances of 1 node on each side
+	 * (chunksize * MAP_BLOCKSIZE + 2)^3 = 551368
+	*/
+	if (cube.X * cube.Y * cube.Z > 551368) {
+		luaL_error(L, "find_nodes_in_area(): area volume"
+				" exceeds allowed value of 551368");
+		return 0;
+	}
+
 	std::set<content_t> filter;
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: unordered_set

File: src/script/lua_api/l_env.cpp [hunk 4]
```diff
@@ -635,35 +635,46 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
...
+		v3s16 p(x, y, z);
+		content_t c = env->getMap().getNodeNoEx(p).getContent();
+		if (filter.count(c) != 0) {
+			push_v3s16(L, p);
+			lua_rawseti(L, -2, ++i);
+			individual_count[c]++;
+		}
 	}
 	lua_newtable(L);
 	for (std::set<content_t>::iterator it = filter.begin();
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: const_iterator

File: src/script/lua_api/l_env.cpp (hunk 3 above)
Comment (SmallJoker) [inline src/script/lua_api/l_env.cpp]: Not possible. `INodeDefManager::getIds` (used below) only takes a "regular" set. Changing that is out of the scope of this PR.

Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: okay then

File: src/script/lua_api/l_env.cpp [hunk 5]
```diff
@@ -651,38 +651,49 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
 	INodeDefManager *ndef = getServer(L)->ndef();
 	v3s16 minp = read_v3s16(L, 1);
 	v3s16 maxp = read_v3s16(L, 2);
+	v3s16 cube = maxp - minp + 1;
+
+	/* Limit for too large areas, assume default values
+	 * and give tolerances of 1 node on each side
+	 * (chunksize * MAP_BLOCKSIZE + 2)^3 = 551368
+	*/
+	if (cube.X * cube.Y * cube.Z > 551368) {
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: are you sure cube.X * cube.Y * cube.Z cannot overflow ? s16 * s16 * s16 = s16 => overflow, in think it's better to have a bigger integer ton contain this result

File: src/script/lua_api/l_env.cpp [hunk 6]
```diff
@@ -651,38 +651,49 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
 	INodeDefManager *ndef = getServer(L)->ndef();
 	v3s16 minp = read_v3s16(L, 1);
 	v3s16 maxp = read_v3s16(L, 2);
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: please prevent underflow in the next calculation (ordering max and min or verify it's correctly ordered)

File: src/script/lua_api/l_env.cpp [hunk 7]
```diff
@@ -706,12 +717,23 @@ int ModApiEnvMod::l_find_nodes_in_area_under_air(lua_State *L)
...
+	v3s16 cube = maxp - minp + 1;
+
+	/* Limit for too large areas, assume default values
+	 * and give tolerances of 1 node on each side
+	 * (chunksize * MAP_BLOCKSIZE + 2)^3 = 551368
+	*/
+	if (cube.X * cube.Y * cube.Z > 551368) {
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: same problem on both cases here

File: src/script/lua_api/l_env.cpp [hunk 8]
```diff
@@ -651,38 +651,51 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
...
 			// removes value, keeps key for next iteration
 			lua_pop(L, 1);
 		}
-	} else if(lua_isstring(L, 3)) {
+	} else if (lua_isstring(L, 3)) {
 		ndef->getIds(lua_tostring(L, 3), filter);
 	}
 
-	std::map<content_t, u16> individual_count;
+	UNORDERED_MAP<content_t, u32> individual_count;
```
Comment (paramat) [inline src/script/lua_api/l_env.cpp]: Do we use 'UNORDERED_MAP' anymore? see C++11 changes.

File: src/script/lua_api/l_env.cpp [hunk 9]
```diff
@@ -651,38 +651,51 @@ int ModApiEnvMod::l_find_nodes_in_area(lua_State *L)
 	INodeDefManager *ndef = getServer(L)->ndef();
 	v3s16 minp = read_v3s16(L, 1);
 	v3s16 maxp = read_v3s16(L, 2);
+	sortBoxVerticies(minp, maxp);
+
+	v3s16 cube = maxp - minp + 1;
+
+	/* Limit for too large areas, assume default values
+	 * and give tolerances of 1 node on each side
+	 * (chunksize * MAP_BLOCKSIZE + 2)^3 = 551368
```
Comment (paramat) [inline src/script/lua_api/l_env.cpp]: `chunksize` varies, should the limit vary? A constant value seems better.
Chunksize 1 would limit this function to a 18^3 nodee volume which is small.

File: src/script/lua_api/l_env.cpp [hunk 10]
```diff
@@ -706,12 +719,25 @@ int ModApiEnvMod::l_find_nodes_in_area_under_air(lua_State *L)
...
```
Comment (paramat) [inline src/script/lua_api/l_env.cpp]: As above.

File: src/script/lua_api/l_env.cpp (hunk 8 above)
Comment (SmallJoker) [inline src/script/lua_api/l_env.cpp]: apparently I forgot to push this change. I even rebased to the newest HEAD in order to fix it properly.

File: src/script/lua_api/l_env.cpp (hunk 6 above)
Comment (paramat) [inline src/script/lua_api/l_env.cpp]: Is this comment addressed?

Comment (SmallJoker) [inline src/script/lua_api/l_env.cpp]: Yes, fixed this issue.

**Comment by nerzhul** [comment]
@SmallJoker please merge if it's ready :)
